ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,184 @@
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that none of those transactions are invalid.
+prop_Mempool_InvalidTxsNeverAdded :: Chain TestBlock
+                                  -> [TestTx]
+                                  -> Property
+prop_Mempool_InvalidTxsNeverAdded bc txs = withMaxSuccess 1000 $
+  tabulate "Transactions" (map constrName txs) $
+  runSimOrThrow test
+ where
+  genTxIsInvalid :: GenTx TestBlock -> Bool
+  genTxIsInvalid (TestGenTx (InvalidTestTx _)) = True
+  genTxIsInvalid _                             = False
+  --
+  test :: forall s. SimM s Property
+  test = withOpenMempool bc $ \mempool -> do
+    let genTxs = testTxsToGenTxPairs txs
+        Mempool
+          { addTxs
+          , getSnapshot
+          } = mempool
+    invalidTxs <- addTxs genTxs
+    MempoolSnapshot {getTxs} <- atomically getSnapshot
+    pure $
+      counterexampleMempoolInfo genTxs invalidTxs getTxs $
+      classifyMempoolSize getTxs $
+      any (\(_, tx, _) -> genTxIsInvalid tx) getTxs === False
+
+{-------------------------------------------------------------------------------
+  Helpers
+-------------------------------------------------------------------------------}
+
+testTxsToGenTxPairs :: [TestTx] -> [(GenTxId TestBlock, GenTx TestBlock)]
+testTxsToGenTxPairs = map testTxToGenTxPair
+
+testTxToGenTxPair :: TestTx -> (GenTxId TestBlock, GenTx TestBlock)
+testTxToGenTxPair tx = (computeGenTxId genTx, genTx)
+ where
+  genTx = TestGenTx tx
+
+openDB :: forall s. SimM s (ChainDB (SimM s) TestBlock)
+openDB = Mock.openDB singleNodeTestConfig testInitExtLedger
+
+withOpenMempool :: Chain TestBlock
+                -> (Mempool (SimM s) TestBlock TicketNo -> SimM s a)
+                -> SimM s a
+withOpenMempool bc action = do
+  chainDb <- ChainDB.fromChain openDB bc
+  withThreadRegistry $ \registry -> do
+    let cfg = ledgerConfigView singleNodeTestConfig
+    action =<< openMempool registry chainDb cfg
+
+-- | Classify whether we're testing against an empty or non-empty 'Mempool' in
+-- each test case.
+classifyMempoolSize :: Testable prop => [a] -> prop -> Property
+classifyMempoolSize txs prop =
+  classify (null txs)         "empty Mempool"     $
+  classify (not . null $ txs) "non-empty Mempool" $
+  prop
+
+counterexampleMempoolInfo :: ( Show a
+                             , Show b
+                             , Show c
+                             , Testable prop
+                             )
+                          => [a]
+                          -> [b]
+                          -> [c]
+                          -> prop
+                          -> Property
+counterexampleMempoolInfo txsToAdd invalidTxs txsInMempool prop =
+  counterexample
+  ("Transactions to add: "         <> show txsToAdd   <> "\n" <>
+   "Invalid transactions: "        <> show invalidTxs <> "\n" <>
+   "Transactions in the Mempool: " <> show txsInMempool) $
+  prop
+
+constrName :: Show a => a -> String
```

Comment:
We use some generics-sop code for this in other tests, but I'm planning to simplify that, so this can stay as-is for now (nice and simple!).

Reply:
If there's an issue for it, I can leave a comment referencing this so you can remember that it exists :slightly_smiling_face:

Reply:
Good idea: #709

Reply:
Don't forget to leave the comment :grin: